metrics_sdk/lib/opentelemetry/sdk/metrics/meter_provider.rb

```diff
@@ -0,0 +1,134 @@
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def shutdown(timeout: nil)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('calling MetricProvider#shutdown multiple times.')
+              Export::FAILURE
+            else
+              start_time = OpenTelemetry::Common::Utilities.timeout_timestamp
+              results = @metric_readers.map do |metric_reader|
+                remaining_timeout = OpenTelemetry::Common::Utilities.maybe_timeout(timeout, start_time)
+                if remaining_timeout&.zero?
+                  Export::TIMEOUT
+                else
+                  metric_reader.shutdown(timeout: remaining_timeout)
+                end
+              end
+
+              @stopped = true
+              results.max || Export::SUCCESS
+            end
+          end
+        end
+
+        # This method provides a way for provider to notify the registered
+        # {MetricReader} instances, so they can do as much as they could to consume
+        # or send the metrics. Note: unlike Push Metric Exporter which can send data on
+        # its own schedule, Pull Metric Exporter can only send the data when it is
+        # being asked by the scraper, so ForceFlush would not make much sense.
+        #
+        # @param [optional Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def force_flush(timeout: nil)
+          @mutex.synchronize do
+            if @stopped
+              Export::SUCCESS
+            else
+              start_time = OpenTelemetry::Common::Utilities.timeout_timestamp
+              results = @metric_readers.map do |metric_reader|
+                remaining_timeout = OpenTelemetry::Common::Utilities.maybe_timeout(timeout, start_time)
+                if remaining_timeout&.zero?
+                  Export::TIMEOUT
+                else
+                  metric_reader.force_flush(timeout: remaining_timeout)
+                end
+              end
+
+              results.max || Export::SUCCESS
+            end
+          end
+        end
+
+        # Adds a new MetricReader to this {MeterProvider}.
+        #
+        # @param metric_reader the new MetricReader to be added.
+        def add_metric_reader(metric_reader)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('calling MetricProvider#add_metric_reader after shutdown.')
+            else
+              @metric_readers.push(metric_reader)
+              @meter_registry.each do |_k, meter|
+                meter.instrument_registry.each do |_n, instrument|
+                  instrument.register_with_new_metric_store(metric_reader.metric_store)
+                end
+              end
```

Comment:
I'd probably move the `meter.instrument_registry ...` bit inside `Meter`, like:
```suggestion
              @meter_registry.each_value { |meter| meter.add_metric_reader(metric_reader) }
```